Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,334 @@
+#ifndef SPIRE_ASSOCIATIVE_VALUE_MODEL_HPP
+#define SPIRE_ASSOCIATIVE_VALUE_MODEL_HPP
+#include <boost/signals2/shared_connection_block.hpp>
+#include "Spire/Ui/Ui.hpp"
+#include "Spire/Ui/LocalValueModel.hpp"
+
+namespace Spire {
+namespace Details {
+  struct AssociatedBooleanModel {
+    std::shared_ptr<BooleanModel> m_model;
+    boost::signals2::scoped_connection m_connection;
+  };
+}
+
+  /** 
+  * Associates BooleanModels with corresponding values, and ensures that
+  * only a single model has a value of true at one time.
+  */
+  template <typename T>
+  class AssociativeValueModel : public ValueModel<T> {
+    public:
+
+      /** Constructs an AssociativeValueModel. */
+      AssociativeValueModel();
+
+      /**
+      * Associates a BooleanModel iff it's not already associated.
+      * If there are no existing models the BooleanModel is set to true,
+      * otherwise, it's set to false.
+      * @param model The model to associate.
+      * @param value The associated value.
+      */
+      void associate(const std::shared_ptr<BooleanModel>& model,
+        const T& value);
+
+      /** Constructs a LocalBooleanModel and associates it with the given
+      * value.
+      * @param value The associated value.
+      * @return The associated model.
+      */
+      std::shared_ptr<BooleanModel> make_association(const T& value);
+
+      /**
+      * Finds the model associated with a value.
+      * @param value The model's associated value.
+      * @return The associated model, or nullptr if no associated model was
+      *         found.
+      */
+      std::shared_ptr<BooleanModel> find(const T& value) const;
+
+      /**
+      * Returns Acceptable if the model's current value is valid, Invalid
+      * otherwise.
+      * @return The state of the AssociativeValueModel.
+      */
+      QValidator::State get_state() const;
+
+      /**
+      * Returns the AssociativeValueModel's current value.
+      * @return The current value.
+      */
+      const T& get_current() const;
+
+      /**
+      * Sets the current value, iff a model is associated with the value.
+      * @param value The current value.
+      * @return Acceptable if the value was set successfully, Invalid
+      *         otherwise.
+      */
+      QValidator::State set_current(const T& value);
+
+      /** Connects a slot to the current signal. */
+      boost::signals2::connection connect_current_signal(
+        const typename CurrentSignal::slot_type& slot) const;
+
+    private:
+      mutable CurrentSignal m_current_signal;
+      T m_current;
```

Comment:
Reference the current value through it's model.

Reply:
The BooleanModels used by the associative model don't have a 'value' in this sense, aside from the boolean value. I can remove m_current and search through the models to find which one has a value of true, but I lose the benefit of having that value -> model mapping.

Reply:
Yep keep it as is then, good point.